File: src/main/java/org/luxons/sevenwonders/game/data/serializers/ResourceTypeSerializer.java
package org.luxons.sevenwonders.game.data.serializers;

import java.lang.reflect.Type;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;
import com.google.gson.JsonPrimitive;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;
import org.luxons.sevenwonders.game.resources.ResourceType;

public class ResourceTypeSerializer implements JsonSerializer<ResourceType>, JsonDeserializer<ResourceType> {

    @Override
    public JsonElement serialize(ResourceType type, Type typeOfSrc, JsonSerializationContext context) {
        return new JsonPrimitive(type.getSymbol());
    }

    @Override
    public ResourceType deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws
            JsonParseException {
        String str = json.getAsString();
        if (str.isEmpty()) {
            throw new IllegalArgumentException("Empty string is not a valid resource type");
        }
        return ResourceType.fromSymbol(str.charAt(0));
    }
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/data/serializers/ResourceTypeSerializer.kt
package org.luxons.sevenwonders.game.data.serializers

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import com.google.gson.JsonPrimitive
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import org.luxons.sevenwonders.game.resources.ResourceType
import java.lang.reflect.Type

internal class ResourceTypeSerializer : JsonSerializer<ResourceType>, JsonDeserializer<ResourceType> {

    override fun serialize(type: ResourceType, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        return JsonPrimitive(type.symbol!!)
    }

    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): ResourceType {
        val str = json.asString
        if (str.isEmpty()) {
            throw IllegalArgumentException("Empty string is not a valid resource type")
        }
        return ResourceType.fromSymbol(str[0])
    }
}
